# Lab 13 - Statistical Tests
## One-Way ANOVA, Post-Hoc Analysis and Friedman Test

Name: Bolohan Sebastian, Lefter Sorin
Group: MIAO2

The dataset contains three groups:

- Group 1 = Control Group
- Group 2 = Mnemonic A
- Group 3 = Mnemonic B

The score represents the number of words recalled.

We want to determine whether the mean scores of the three groups are significantly different.

H0: μ1 = μ2 = μ3

The means of all groups are equal.

H1:

At least one group mean differs from the others.

DATA

In [1]:
group1 = [3,5,3,2,4,6,9,3,8,10]

group2 = [10,8,15,9,11,16,17,15,7,10]

group3 = [29,15,14,15,17,10,8,11,18,19]

DESCRIPTIVE STATISTICS

In [2]:
import numpy as np

print("Group 1 mean:", np.mean(group1))
print("Group 2 mean:", np.mean(group2))
print("Group 3 mean:", np.mean(group3))

Group 1 mean: 5.3
Group 2 mean: 11.8
Group 3 mean: 15.6


ONE WAY ANOVA

In [4]:
from scipy.stats import f_oneway

F, p = f_oneway(group1, group2, group3)

print("F statistic =", F)
print("p-value =", p)
alpha = 0.05

if p < alpha:
    print("Reject H0")
else:
    print("Fail to reject H0")

F statistic = 14.58892650866362
p-value = 5.06152585965283e-05
Reject H0


If p < 0.05, the mean scores are significantly different.

TURKEY POST-HOC


In [5]:
import numpy as np
from statsmodels.stats.multicomp import pairwise_tukeyhsd

values = np.array(
    group1 +
    group2 +
    group3
)

groups = (
    ['Control'] * len(group1) +
    ['MnemonicA'] * len(group2) +
    ['MnemonicB'] * len(group3)
)

tukey = pairwise_tukeyhsd(
    values,
    groups,
    alpha=0.05
)

print(tukey)

   Multiple Comparison of Means - Tukey HSD, FWER=0.05    
  group1    group2  meandiff p-adj   lower   upper  reject
----------------------------------------------------------
  Control MnemonicA      6.5 0.0062  1.7183 11.2817   True
  Control MnemonicB     10.3    0.0  5.5183 15.0817   True
MnemonicA MnemonicB      3.8  0.139 -0.9817  8.5817  False
----------------------------------------------------------


The Tukey test identifies which pairs of groups differ significantly.
A pair is significantly different when reject=True.

SCHEFFE TEST

In [6]:
!pip install scikit-posthocs

In [8]:
import pandas as pd
import scikit_posthocs as sp

df = pd.DataFrame({
    'group':
        ['Control']*10 +
        ['MnemonicA']*10 +
        ['MnemonicB']*10,
    'score':
        group1 + group2 + group3
})
sp.posthoc_scheffe(
    df,
    val_col='score',
    group_col='group'
)

,Control,MnemonicA,MnemonicB
Control,1.000000,0.008731,0.000059
MnemonicA,0.008731,1.000000,0.163040
MnemonicB,0.000059,0.163040,1.000000


Pairs with p-value < 0.05 are significantly different.

Exercise 2 - Friedman Test

HYPOTHESIS:

H0:

All algorithms perform equally.

H1:

At least one algorithm performs differently.

In [9]:
from scipy.stats import friedmanchisquare

mh1 = [
565.02,662.84,664.73,857.84,949.98,
1084.82,837.80,906.16,1000.27,1076.88,
1170.17,1217.01,1364.50,1464.20,1544.21,
1064.89,1104.67,1202.00,887.22,963.06,
952.29
]

mh2 = [
567.98,619.35,629.59,809.13,858.98,
949.89,832.91,881.26,955.95,1052.65,
1107.47,1184.58,1296.33,1384.13,1488.71,
1003.00,1042.79,1141.94,813.98,852.89,
914.04
]

mh3 = [
568.86,617.48,620.50,817.71,858.95,
942.60,838.50,882.70,921.97,1074.38,
1108.88,1166.59,1340.98,1367.91,1454.91,
1007.26,1035.23,1110.13,823.01,859.06,
915.38
]

mh4 = [
565.99,614.23,618.04,803.51,841.63,
961.47,830.48,876.21,918.45,1050.11,
1100.95,1158.88,1305.83,1354.04,1437.52,
1003.07,1042.61,1118.63,819.81,860.12,
909.06
]

stat, p = friedmanchisquare(
    mh1,
    mh2,
    mh3,
    mh4
)

print("Friedman statistic =", stat)
print("p-value =", p)

Friedman statistic = 35.457142857142856
p-value = 9.753018638534493e-08


The Friedman test was applied to compare the performance of four metaheuristics over 21 benchmark instances.

The obtained Friedman statistic was:

χ² = 35.457

The corresponding p-value was:

p = 9.75 × 10⁻⁸

Since p < 0.05, the null hypothesis is rejected.

Therefore, there is a statistically significant difference between the performances of the four metaheuristics.

This indicates that at least one metaheuristic performs differently from the others across the benchmark instances.